server/pulp/server/controllers/distributor.py

```diff
@@ -1,109 +1,294 @@
+    async_result = update.apply_async_with_reservation(
+        tags.RESOURCE_REPOSITORY_TYPE, distributor.repo_id,
+        [distributor, config, delta],
+        tags=task_tags)
+    return async_result
+
+
 @celery.task(base=Task, name='pulp.server.tasks.repository.distributor_update')
-def update(repo_id, distributor_id, config, delta):
+def update(distributor, config, delta):
     """
-    Get the itinerary for updating a repository distributor.
-      1. Update the distributor on the server.
-      2. (re)bind any bound consumers.
-
-    :param repo_id:         A repository ID.
-    :type  repo_id:         str
-    :param distributor_id:  A unique distributor id
-    :type  distributor_id:  str
-    :param config:          A configuration dictionary for a distributor instance. The contents of
-                            this dict depends on the type of distributor.
-    :type  config:          dict
-    :param delta:           A dictionary used to change other saved configuration values for a
-                            distributor instance. This currently only supports the 'auto_publish'
-                            keyword, which should have a value of type bool
-    :type  delta:           dict or None
-
-    :return: Any errors that may have occurred and the list of tasks spawned for each consumer
-    :rtype: TaskResult
+    Update the distributor and (re)bind any bound consumers.
+
+    :param distributor: distributor to be updated
+    :type  distributor: pulp.server.db.model.Distributor
+    :param config: A configuration dictionary for a distributor instance. The contents of this dict
+                   depends on the type of distributor. Values of None will remove they key from the
+                   config. Keys ommited from this dictionary will remain unchanged.
+    :type  config: dict
+    :param delta: A dictionary used to change conf values for a distributor instance. This currently
+                  only supports the 'auto_publish' keyword, which should have a value of type bool
+    :type  delta: dict or None
+
+    :return: result containing any errors and tasks spawned
+    :rtype pulp.server.async.tasks.TaskResult
     """
-    manager = managers.repo_distributor_manager()
+    repo = model.Repository.objects.get_repo_or_missing_resource(distributor.repo_id)
 
-    # Retrieve configuration options from the delta
-    auto_publish = None
-    if delta is not None:
-        auto_publish = delta.get('auto_publish')
+    for k, v in config.iteritems():
+        if v is None:
+            distributor.config.pop(k)
+        else:
+            distributor.config[k] = v
 
-    distributor = manager.update_distributor_config(repo_id, distributor_id, config, auto_publish)
+    auto_publish = delta.get('auto_publish') if delta else None
+    if not isinstance(auto_publish, (bool, type(None))):
+        raise exceptions.InvalidValue(['auto_publish'])
+    distributor.auto_publish = auto_publish or distributor.auto_publish
```

Comment:
If ``distributor.auto_publish == True`` and ``delta['auto_publish'] == False``, I think this will neglect to set the value to ``False`` on the object.